lib/src/rules/use_enums.dart

```diff
@@ -0,0 +1,244 @@
+    initialization expression of these static fields,
+  * do not define `hashCode`, `==` or `index`,
+  * do not extend any class other than Object, and
+  * have no subclasses declared in the defining library.
+
+**BAD:**
+```dart
+class LogPriority {
+  static const error = LogPriority._(1, 'Error');
+  static const warning = LogPriority._(2, 'Warning');
+  static const unknown = LogPriority._unknown('Log');
+
+  final String prefix;
+  final int priorty;
+  const LogPriority._(this.priority, this.prefix);
+  const LogPriority._unknown(String prefix) : this._(-1, prefix);
+}
+```
+
+**GOOD:**
+```dart
+enum LogPriority {
+  error(1, 'Error'),
+  warning(2, 'Warning'),
+  log.unknown('Log');
+ 
+  final String prefix;
+  final int priorty;
+  LogPriority(this.priority, this.prefix);
+  LogPriority.unknown(String prefix) : this(-1, prefix);
+}
+```
+''';
+
+class UseEnums extends LintRule {
+  UseEnums()
+      : super(
+            name: 'use_enums',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    if (!context.isEnabled(Feature.enhanced_enums)) return;
+
+    var visitor = _Visitor(this, context);
+    registry.addClassDeclaration(this, visitor);
+  }
+}
+
+/// A superclass for the [_EnumVisitor] and [_NonEnumVisitor].
+class _BaseVisitor extends RecursiveAstVisitor<void> {
+  /// The element representing the enum declaration that's being visited.
+  final ClassElement classElement;
+
+  _BaseVisitor(this.classElement);
+
+  /// Return `true` if the given [node] is an invocation of a generative
+  /// constructor from the class being converted.
+  bool invokesGenerativeConstructor(InstanceCreationExpression node) {
+    var constructorElement = node.constructorName.staticElement;
+    return constructorElement != null &&
+        !constructorElement.isFactory &&
+        constructorElement.enclosingElement == classElement;
+  }
+}
+
+/// A visitor used to visit the class being linted. This visitor throws an
+/// exception if a constructor for the class is invoked anywhere other than the
+/// top-level expression of an initializer for one of the fields being converted.
+///
+/// Note: copied and modified from server.
+class _EnumVisitor extends _BaseVisitor {
+  /// A flag indicating whether we are currently visiting the children of a
+  /// field declaration that will be converted to be a constant.
+  bool inConstantDeclaration = false;
+
+  List<VariableDeclaration> variableDeclarations;
+
+  _EnumVisitor(ClassElement classElement, this.variableDeclarations)
+      : super(classElement);
+
+  @override
+  void visitInstanceCreationExpression(InstanceCreationExpression node) {
+    if (!inConstantDeclaration) {
+      if (invokesGenerativeConstructor(node)) {
+        throw _InvalidEnumException();
+      }
+    }
+    inConstantDeclaration = false;
```

Comment:
I can't figure out how to respond to the existing thread.

You're not missing anything, I was. Because we're visiting a `VariableDeclaration` rather than the enclosing `FieldDeclaration`, the flag is getting reset for every field. Sorry for the confusion.